pkg/controller/publishingstrategy/publishingstrategy_controller.go

```diff
@@ -105,80 +105,50 @@ func (r *ReconcilePublishingStrategy) Reconcile(request reconcile.Request) (reco
+	for _, ingressDefinition := range instance.Spec.ApplicationIngress {
 
-	// create list of applicationIngress
-	var ingressNotOnCluster []cloudingressv1alpha1.ApplicationIngress
-
-	exisitingIngressMap := convertIngressControllerToMap(ingressControllerList.Items)
-
-	// loop through every applicationingress in publishing strategy
-	for _, publishingStrategyIngress := range instance.Spec.ApplicationIngress {
-		if !checkExistingIngress(exisitingIngressMap, &publishingStrategyIngress) {
-			ingressNotOnCluster = append(ingressNotOnCluster, publishingStrategyIngress)
+		// Set the namespaced name
+		namespacedName := types.NamespacedName{Name: getIngressName(ingressDefinition.DNSName), Namespace: ingressControllerNamespace}
+		// Default
+		if ingressDefinition.Default {
+			namespacedName = types.NamespacedName{Name: "default", Namespace: ingressControllerNamespace}
 		}
-	}
-
-	// ----------------------------TODO: remove these debug logs afterwards --------------------------------------
-	var serializedIngressControllerList bytes.Buffer
-	_ = serializer.Encode(ingressControllerList, &serializedIngressControllerList)
-	log.Info(fmt.Sprintf("initial ingresscontroller list: %s", serializedIngressControllerList.String()))
-	log.Info(fmt.Sprintf("appingress on publishingstrategy CR: %+v", instance.Spec.ApplicationIngress))
-	log.Info(fmt.Sprintf("ingress to reconcile: %+v", ingressNotOnCluster))
-	// -----------------------------------------------------------------------------------------------------------
 
-	for _, appingress := range ingressNotOnCluster {
-		newCertificate := &corev1.LocalObjectReference{
-			Name: appingress.Certificate.Name,
+		desiredIngressController := generateIngressController(ingressDefinition)
+		// Try to get matching IngressController
+		ingressController := &operatorv1.IngressController{}
+		err = r.client.Get(context.TODO(), namespacedName, ingressController)
+		if err != nil {
+			// Attempt to create the CR if not found
+			if k8serr.IsNotFound(err) {
+				err = r.client.Create(context.TODO(), desiredIngressController)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+				// If the CR was created, requeue PublishingStrategy
+				return reconcile.Result{Requeue: true}, nil
+			}
+			return reconcile.Result{}, err
 		}
 
-		if appingress.Default == true {
-			err := r.defaultIngressHandle(appingress, ingressControllerList, newCertificate)
+		// If the IngressController exists, ensure it matches PublishingStrategies definition
+		if !reflect.DeepEqual(ingressController.Spec, desiredIngressController.Spec) {
+			// Deafult is a special case, the matching data can also be in the status
+			if ingressDefinition.Default {
+				if checkDefaultIngressControllerStatus(*ingressController, desiredIngressController.Spec) {
+					continue
+				}
+			}
+			// If the definitions aren't the same, delete the IngressController
+			err = r.client.Delete(context.TODO(), ingressController)
```

Comment:
we only want to delete if the dns and/or listening(internal/external) change

Reply:
Is there a desgin doc for this? OSD-5507 mentions using DeepEqual

Reply:
I made a comment on the card I believe. No design doc, but we shouldn't delete if we can patch imo because delete leads to dns blips which is nice to avoid

Reply:
+1 to Sam. We should only delete if it's an immutable field (domain and endpointpublishingstrategy I think are the only two).